### PR TITLE
[New Op] Added dropout unary op

### PR DESCRIPTION
**Pull request by dmakoviichuk-tt**
### Ticket
None

### Problem description
I've found that we have lkk kernel but no unary dropout.
### What's changed
Added unary op: dropout.
Added usage example.

Unfortunately reproducibility is low. I was able to get same results twice only if I reset device between calls.
I implemented using_dropout.py example but didn't add tests because I cannot get the same result twice.
Feedback to the kernel team was provided.
Also std is pretty high but I can use it.
Comparison vs pytorch:
```
TTNN:
Expected probability: 0.2
Mean: 0.1978125
Standard Deviation: 0.1085284200969958
PYTORCH:
Expected probability: 0.2
Mean: 0.199869384765625
Standard Deviation: 0.002019949620664387
```

https://github.com/tenstorrent/tt-metal/actions/runs/10874629279
### Checklist
- [x] Post commit CI passes
- [x] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp [hunk 1]
```diff
@@ -278,6 +290,39 @@ Tensor Softplus::invoke(
         optional_output_tensor);
 }
 
+Tensor Dropout::invoke(
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:293]: Wdyt about extracting this to its own .hpp/cpp?

**Review by ayerofieiev-tt — COMMENTED** [review]
Overall LGTM, but format change makes it hard to distinguish actual change from everything else. I think it is best to revert style changes.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp (hunk 1 above)
Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:308]: Don't want to do it right now. It is implemented the same way as any other op and uses its infra.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp [hunk 2]
```diff
@@ -1233,6 +1234,60 @@ void bind_unary_operation_with_diag(py::module& module, const unary_operation_t&
...
+                * :attr:`probability` (float): Dropout probability. In average total_elems * probability elements will be zero out.
+                * :attr:`scale` (float): Scales output tensor. In general scale == 1.0/(1.0-probability)
+                * :attr:`memory_config` (Optional[ttnn.MemoryConfig]): Memory configuration for the operation.
+                * :attr:`output_tensor` (Optional[ttnn.Tensor]): preallocated output tensor
+                * :attr:`queue_id` (Optional[uint8]): command queue id
+
+            Example:
+
+                >>> tensor = ttnn.from_torch(torch.tensor((1, 2), dtype=torch.bfloat16), device=device)
+                >>> output = {1}(tensor, seed=228, probability=0.2, scale= 1.0/(1.0 - probability))
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp:1261]: @rfurko-tt  here is example of real usage.

File: ttnn/examples/usage/using_dropout.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            32,
+        )
+    )
+    t_tt = ttnn.from_torch(t, device=device, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT)
+
+    ratios = []
+    s = 124
+    prob = 0.2
+    for _ in range(100):
+        output = ttnn.dropout(t_tt, seed=s, probability=prob, scale=1.0)
```
Comment (rfurko-tt) [inline ttnn/examples/usage/using_dropout.py:34]: Discussed offline. My opinion that we need to change scale to 1/prob and change output = dropout(t) (without * (1 - prob)). Feel free to choose any option.

**Review by rfurko-tt — APPROVED** [review]
LGTM

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp [hunk 4]
```diff
@@ -1233,6 +1234,60 @@ void bind_unary_operation_with_diag(py::module& module, const unary_operation_t&
...
+               const uint32_t seed,
+               const float probability,
+               const float scale,
+               const std::optional<MemoryConfig>& memory_config,
+               const std::optional<Tensor>& output_tensor,
+               const uint8_t queue_id) {
+                return self(queue_id, input, seed, probability, scale, memory_config, output_tensor);
+            },
+            py::arg("input_tensor"),
+            py::kw_only(),
```
Comment (cfjchu) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp:1282]: 👍

**Comment by cfjchu** [comment]
who's taking a look on the kernel team?

**Review by cfjchu — APPROVED** [review]

**Comment by dmakoviichuk-tt** [comment]
@eyonland @yan-zaretskiy  please review

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_types.hpp
Comment (amahmudTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_types.hpp:1]: As we discussed earlier it is probably initializing the seed for every tile, which will give bad random number performance. Since that is how all the other standard ops work afaik. The seed needs to be initialized once for all the tiles.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp [hunk 5]
```diff
@@ -305,6 +305,39 @@ Tensor Identity::invoke(
...
+    const uint32_t seed,
+    const float probability,
+    const float scale,
+    const std::optional<MemoryConfig>& memory_config,
+    const std::optional<Tensor>& optional_output_tensor) {
+    TT_ASSERT(input.device()->arch() != tt::ARCH::GRAYSKULL, "Dropout is not currently supported on Grayskull");
+    return detail::unary_impl(
+        queue_id,
+        input,
+        {UnaryWithParam{UnaryOpType::DROPOUT, {static_cast<float>(seed), probability, scale}}},
```
Comment (amahmudTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:336]: Same question as others, why is the seed in float ?

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp [hunk 6]
```diff
@@ -1233,6 +1234,60 @@ void bind_unary_operation_with_diag(py::module& module, const unary_operation_t&
...
+            Applies {0} to :attr:`input_tensor` element-wise.
+
+            .. math::
+                {0}(\\mathrm{{input\\_tensor}}_i)
+
+            Args:
+                * :attr:`input_tensor`
+
+            Keyword Args:
+                * :attr:`seed` (float): seed used for RNG
```
Comment (amahmudTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp:1251]: Why float, but if it is float then what about line 1240 ?

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp (hunk 5 above)
Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:336]: It is the limitation of the current unary op implementation. They are using vector<float> to pass parameters.
I cannot do much here.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp (hunk 6 above)
Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary_pybind.hpp:1251]: Good catch, will fix it.

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp (hunk 5 above)
Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:336]: In my opinion all these ops should be rewritten.  But it is out of scope of the current pr.

File: ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_types.hpp
Comment (dmakoviichuk-tt) [reply ttnn/cpp/ttnn/operations/eltwise/unary/common/unary_op_types.hpp:1]: Yes, problem is that currently it requires to fully rewrite unary op implementation to support new features and remove casting to the std::vector<float> of all input parameters.
I am going to do it but in other pr.

File: ttnn/examples/usage/using_dropout.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+# SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+
+
+import ttnn
```
Comment (ayerofieiev-tt) [inline ttnn/examples/usage/using_dropout.py:6]: If this is not going to be tested, this should be removed.
If it is important to have it, make sure this code is run on CI

Comment (dmakoviichuk-tt) [reply ttnn/examples/usage/using_dropout.py:6]: do you want me to remove example of using_dropout?
It is nice to have it to quickly check in future.

Comment (eyonland) [reply ttnn/examples/usage/using_dropout.py:6]: It does not hurt to have an example.   Can we also have a unit test for this functionality?

File: ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp (hunk 5 above)
Comment (eyonland) [reply ttnn/cpp/ttnn/operations/eltwise/unary/unary.cpp:336]: I do agree with you here.  Unfortunately we have a freeze on the API at the moment unless this blocks models or customers.

**Review by eyonland — CHANGES_REQUESTED** [review]
Please add a unit test in addition to the example.

**Comment by dmakoviichuk-tt** [comment]
@eyonland  I explained in the description why I cannot have unit test because of the reproducibility issues.
Who said that we need to freeze on the API?
Your Contractors add more changes to the api day by day.
This dropout blocks my ML library. I think this is enough.

**Review by eyonland — APPROVED** [review]
